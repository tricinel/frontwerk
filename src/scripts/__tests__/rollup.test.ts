/* eslint-disable global-require */

import cases from 'jest-in-case';
import jestSerializerPath from 'jest-serializer-path';

import {
  unquoteSerializer,
  winPathSerializer
} from '../../helpers/serializers';

jest.mock('concurrently');
jest.mock('rimraf');
jest.mock('consola');

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);
expect.addSnapshotSerializer(jestSerializerPath);

const testCases = [
  {
    name: 'calls rollup CLI with default args'
  },
  {
    name: '--no-clean will not clean the dist directory',
    args: ['--no-clean']
  },
  {
    name: 'watches with --watch',
    args: ['--watch']
  },
  {
    name: '--environment will pass the environment',
    args: ['--environment', 'production']
  },
  {
    name: 'does not use built-in config with rollup.config.js file',
    fileExists: (filename: string): boolean => filename === 'rollup.config.js'
  },
  {
    name: 'does not use built-in config with --config',
    args: ['--config', './custom-config.js']
  }
];

const testFn = ({ fileExists = () => false, args = [] }) => {
  const crossSpawn = require('cross-spawn');
  const syncSpy = jest
    .spyOn(crossSpawn, 'sync')
    .mockImplementation(() => ({ status: 0 }));
  const exitSpy = jest.spyOn(process, 'exit');
  const { argv: originalArgv } = process;

  Object.assign(require('../../utils/fileExists'), { default: fileExists });
  Object.assign(require('../../utils/resolveBin'), {
    resolveBin: (modName: string, { executable = modName } = {}) => executable
  });

  try {
    // Tests
    process.argv = ['node', '../rollup', ...args];

    require('../rollup');

    expect(syncSpy).toHaveBeenCalledTimes(1);
    const [firstCall] = syncSpy.mock.calls;
    const [script, calledArgs] = firstCall as [string, string[]];
    expect([script, ...calledArgs].join(' ')).toMatchSnapshot();
  } finally {
    // We reset everything afterEach
    exitSpy.mockRestore();
    syncSpy.mockRestore();
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('rollup', testFn, testCases);
