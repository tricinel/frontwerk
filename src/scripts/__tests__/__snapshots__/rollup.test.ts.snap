// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup --environment will pass the environment 1`] = `concurrently --names esm,cjs,umd,umd.min "cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup --config --environment production" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup --config --environment production" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --config --environment production --sourcemap" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --config --environment production --sourcemap"`;

exports[`rollup --no-clean will not clean the dist directory 1`] = `concurrently --names esm,cjs,umd,umd.min "cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup --config" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup --config" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --config --sourcemap" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --config --sourcemap"`;

exports[`rollup calls rollup CLI with default args 1`] = `concurrently --names esm,cjs,umd,umd.min "cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup --config" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup --config" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --config --sourcemap" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --config --sourcemap"`;

exports[`rollup does not use built-in config with --config 1`] = `concurrently --names esm,cjs,umd,umd.min "cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --sourcemap" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --sourcemap"`;

exports[`rollup does not use built-in config with rollup.config.js file 1`] = `concurrently --names esm,cjs,umd,umd.min "cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup --config" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup --config" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --config --sourcemap" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --config --sourcemap"`;

exports[`rollup watches with --watch 1`] = `concurrently --names esm,cjs,umd,umd.min "cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup --config --watch" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup --config --watch" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --config --watch --sourcemap" "cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --config --watch --sourcemap"`;
